### PR TITLE
Code quality & Add bytecode version 6 support (size_protos 1 byte off moment)

### PR DESCRIPTION
**Pull request by DataM0del**
> [!NOTE]
> This PR still includes bytecode v5 support. It just adds v6 support & cleans up code.
> If you want to test with bytecode v5, use a Luau version after [0.613](https://github.com/luau-lang/luau/releases/tag/0.613) and before [0.630](https://github.com/luau-lang/luau/releases/tag/0.631), the last version with bytecode v5 by default is [0.630](https://github.com/luau-lang/luau/releases/tag/0.630).
> Any recent version of Luau should work with this, as bytecode v6 is supported.

Not all opcodes are implemented in the disassembler. The decompiler has more support, although I think I did implement `FASTCALL3` (what v6 added) support for the disassembler, not sure because it is 10:30 PM CDT & I am typing this on a phone.
Supported opcodes (Luau v6):
- [x] NOP
- [x] BREAK
- [x] LOADNIL
- [x] LOADB
- [x] LOADN
- [x] LOADK
- [x] MOVE
- [x] GETGLOBAL
- [x] SETGLOBAL
- [x] GETUPVAL
- [x] SETUPVAL
- [x] CLOSEUPVALS
- [x] GETIMPORT
- [x] GETTABLE
- [x] SETTABLE
- [x] GETTABLEKS
- [x] SETTABLEKS
- [x] GETTABLEN
- [x] SETTABLEN
- [x] NEWCLOSURE
- [x] NAMECALL
- [x] CALL
- [x] RETURN
- [x] JUMP
- [x] JUMPBACK
- [x] JUMPIF
- [x] JUMPIFNOT
- [x] JUMPIFEQ
- [x] JUMPIFLE
- [x] JUMPIFLT
- [x] JUMPIFNOTEQ
- [x] JUMPIFNOTLE
- [x] JUMPIFNOTLT
- [x] ADD
- [x] SUB
- [x] MUL
- [x] DIV
- [x] MOD
- [x] POW
- [x] ADDK
- [x] SUBK
- [x] MULK
- [x] DIVK
- [x] MODK
- [x] POWK
- [x] AND
- [x] OR
- [x] ANDK
- [x] ORK
- [x] CONCAT
- [x] NOT
- [x] MINUS
- [x] LENGTH
- [x] NEWTABLE
- [x] DUPTABLE
- [x] SETLIST
- [x] FORNPREP
- [x] FORNLOOP
- [x] FORGLOOP
- [x] FORGPREP_INEXT
- [x] FORGPREP_NEXT
- [x] NATIVECALL
- [x] GETVARARGS
- [x] DUPCLOSURE
- [x] PREPVARARGS
- [x] LOADKX
- [x] JUMPX
- [x] FASTCALL
- [x] COVERAGE
- [x] CAPTURE
- [x] SUBRK
- [x] DIVRK
- [x] FASTCALL1
- [x] FASTCALL2
- [x] FASTCALL2K
- [x] FORGPREP
- [x] JUMPXEQKNIL
- [x] JUMPXEQKB
- [x] JUMPXEQKN
- [x] JUMPXEQKS
- [x] IDIV
- [x] IDIVK
- [x] FASTCALL3

### CODE REVIEW

**Comment by focat69** [comment]
fire

**Comment by focat69** [comment]
waiting for ready signal!

**Comment by DataM0del** [comment]
> [!NOTE]
> Some opcodes aren't implemented yet.

**Comment by focat69** [comment]
Bros making my life easy wtf

**Comment by DataM0del** [comment]
It's 10 PM, so I'll probably go to bed in a bit.

**Comment by DataM0del** [comment]
image
ah yes, not adding, exponentially multiply! That's totally what I wanted, I totally didn't want the `ADD` instruction to **add** a number, I wanted it to **exponentially multiply**. Gonna fix it.

**Comment by DataM0del** [comment]
It should be fixed now.

**Comment by DataM0del** [comment]
image
Fixed. Will commit & push this new update.

**Comment by DataM0del** [comment]
I committed my `input.luau` changes again, rebasing to remove them again...

**Comment by DataM0del** [comment]
Going to bed now. Will work on this tomorrow.

**Comment by focat69** [comment]
🤑

**Comment by DataM0del** [comment]
I broke something when I moved the op table to a variable in a different file.

**Comment by DataM0del** [comment]
actually, probably not, because something in here is trying to get the opcode for `65`...

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+	// Use IntelliSense to learn about possible attributes.
+	// Hover to view descriptions of existing attributes.
+	// For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Comment (DataM0del) [inline .vscode/launch.json:4]:
```suggestion
```
Probably should remove this so GitHub doesn't complain about the comments, which is an addition to JSON that only some JSON parsers implement.
Could mark this as a JSONC file with the `.gitattributes`, but I'm on my phone..

**Review by DataM0del — COMMENTED** [review]
One small review because it is 10:37 PM CDT and I am on a phone.

**Comment by focat69** [comment]
almost ready??

**Comment by DataM0del** [comment]
@focat69 I will make a quick script that will list all opcodes I am missing implementations for. Maybe later I will port the handlers to the disassembler.

**Comment by DataM0del** [comment]
Implemented. Here's the results (based on bytecode v6 table):

- [ ] NOP
- [ ] BREAK
- [x] LOADNIL
- [x] LOADB
- [x] LOADN
- [x] LOADK
- [x] MOVE
- [x] GETGLOBAL
- [x] SETGLOBAL
- [x] GETUPVAL
- [x] SETUPVAL
- [x] CLOSEUPVALS
- [x] GETIMPORT
- [x] GETTABLE
- [x] SETTABLE
- [x] GETTABLEKS
- [x] SETTABLEKS
- [x] GETTABLEN
- [x] SETTABLEN
- [x] NEWCLOSURE
- [x] NAMECALL
- [x] CALL
- [x] RETURN
- [x] JUMP
- [x] JUMPBACK
- [x] JUMPIF
- [x] JUMPIFNOT
- [x] JUMPIFEQ
- [x] JUMPIFLE
- [ ] JUMPIFLT
- [ ] JUMPIFNOTEQ
- [x] JUMPIFNOTLE
- [x] JUMPIFNOTLT
- [x] ADD
- [x] SUB
- [x] MUL
- [x] DIV
- [x] MOD
- [x] POW
- [x] ADDK
- [x] SUBK
- [x] MULK
- [x] DIVK
- [x] MODK
- [x] POWK
- [ ] AND
- [ ] OR
- [ ] ANDK
- [ ] ORK
- [ ] CONCAT
- [ ] NOT
- [ ] MINUS
- [ ] LENGTH
- [ ] NEWTABLE
- [ ] DUPTABLE
- [ ] SETLIST
- [x] FORNPREP
- [x] FORNLOOP
- [x] FORGLOOP
- [x] FORGPREP_INEXT
- [x] FORGPREP_NEXT
- [ ] NATIVECALL
- [ ] GETVARARGS
- [ ] DUPCLOSURE
- [x] PREPVARARGS
- [ ] LOADKX
- [x] JUMPX
- [x] FASTCALL
- [x] COVERAGE
- [x] CAPTURE
- [ ] SUBRK
- [ ] DIVRK
- [ ] FASTCALL1
- [ ] FASTCALL2
- [ ] FASTCALL2K
- [x] FORGPREP
- [ ] JUMPXEQKNIL
- [ ] JUMPXEQKB
- [ ] JUMPXEQKN
- [ ] JUMPXEQKS
- [ ] IDIV
- [ ] IDIVK
- [x] FASTCALL3

**Comment by DataM0del** [comment]
4 opcodes left. This was done in 2 weeks (or 1) also.

**Comment by DataM0del** [comment]
Done.

**Comment by DataM0del** [comment]
Actually done now, a better decompiler will be a separate PR.

**Comment by focat69** [comment]
awesome job!

**Pull request merged**